packages/libp2p/src/transport-manager.ts

```diff
@@ -262,27 +257,26 @@ export class DefaultTransportManager implements TransportManager, Startable {
    * If a transport has any running listeners, they will be closed.
    */
   async remove (key: string): Promise<void> {
-    const listeners = this.listeners.get(key) ?? []
-    log.trace('removing transport %s', key)
-
     // Close any running listeners
-    const tasks = []
-    log.trace('closing listeners for %s', key)
-    while (listeners.length > 0) {
-      const listener = listeners.pop()
-
-      if (listener == null) {
-        continue
-      }
-
-      tasks.push(listener.close())
+    for (const listener of this.listeners.get(key)?.values() ?? []) {
+      await listener.close()
     }
-    await Promise.all(tasks)
-
     this.transports.delete(key)
     this.listeners.delete(key)
   }
 
+  async removeListener (ma: Multiaddr): Promise<void> {
+    log('removing listener for %a', ma)
+
+    for (const listeners of this.listeners.values()) {
+      const listener = listeners.get(ma)
+      if (listener != null) {
+        await listener.close()
+        listeners.delete(ma)
```

Comment:
Is there a situation where a single multiaddr can be attached to multiple listeners? I think no.
In that case we can `return` here after removing/closing the first match.